dask/dataframe/shuffle.py

```diff
@@ -1055,7 +1065,7 @@ def compute_divisions(df, col=None, **kwargs) -> tuple:
 def compute_and_set_divisions(df, **kwargs):
     mins, maxes, lens = _compute_partition_stats(df.index, allow_overlap=True, **kwargs)
     if len(mins) == len(df.divisions) - 1:
-        df.divisions = tuple(mins) + (maxes[-1],)
+        df._divisions = tuple(mins) + (maxes[-1],)
```

Comment:
We avoid the `setter` in cases where we are explicitly mutating the structure of `df`. `inplace` methods that may change how partitioning is done need to avoid the `len` check, since the number of partitions may have changed. Everywhere else we should use the `setter`, which is where all the other bugs were caught from.

Reply:
Makes sense